spring-pulsar/src/main/java/org/springframework/pulsar/core/reactive/ReactivePulsarSenderTemplate.java

```diff
@@ -0,0 +1,221 @@
+	 * recording observations
+	 */
+	public ReactivePulsarSenderTemplate(ReactivePulsarSenderFactory<T> reactiveMessageSenderFactory,
+			@Nullable ObservationRegistry observationRegistry,
+			@Nullable PulsarTemplateObservationConvention observationConvention) {
+		this.reactiveMessageSenderFactory = reactiveMessageSenderFactory;
+		this.observationRegistry = observationRegistry;
+		this.observationConvention = observationConvention;
+	}
+
+	@Override
+	public Mono<MessageId> send(T message) {
+		return doSend(null, message, null, null, null);
+	}
+
+	@Override
+	public Mono<MessageId> send(String topic, T message) {
+		return doSend(topic, message, null, null, null);
+	}
+
+	@Override
+	public ReactiveSendMessageBuilderImpl<T> newMessage(T message) {
+		return new ReactiveSendMessageBuilderImpl<>(this, message);
+	}
+
+	@Override
+	public void setBeanName(String beanName) {
+		this.beanName = beanName;
+	}
+
+	/**
+	 * Set the schema to use on this template.
+	 * @param schema provides the {@link Schema} used on this template
+	 */
+	public void setSchema(Schema<T> schema) {
+		this.schema = schema;
+	}
+
+	private Mono<MessageId> doSend(String topic, T message,
+			MessageSpecBuilderCustomizer<T> messageSpecBuilderCustomizer, MessageRouter messageRouter,
+			ReactiveMessageSenderBuilderCustomizer<T> customizer) {
+		final String topicName = ReactiveMessageSenderUtils.resolveTopicName(topic, this.reactiveMessageSenderFactory);
+		this.logger.trace(() -> String.format("Sending reative msg to '%s' topic", topicName));
+
+		PulsarMessageSenderContext senderContext = PulsarMessageSenderContext.newContext(topicName, this.beanName);
+		Observation observation = newObservation(senderContext);
+		try {
+			observation.start();
+			final ReactiveMessageSender<T> sender = createMessageSender(topic, message, messageRouter, customizer);
+
+			MessageSpecBuilder<T> messageSpecBuilder = MessageSpec.builder(message);
+
+			if (messageSpecBuilderCustomizer != null) {
+				messageSpecBuilderCustomizer.customize(messageSpecBuilder);
+			}
+
+			// propagate props to message
+			senderContext.properties().forEach(messageSpecBuilder::property);
+
+			MessageSpec<T> messageSpec = messageSpecBuilder.build();
+			return sender.sendMessage(Mono.just(messageSpec)).doOnError(ex -> {
+				this.logger.error(ex, () -> String.format("Failed to send msg to '%s' topic", topicName));
+				observation.error(ex);
+				observation.stop();
+			}).doOnSuccess(msgId -> {
+				this.logger.trace(() -> String.format("Sent msg to '%s' topic", topicName));
+				observation.stop();
+			});
+		}
+		catch (RuntimeException ex) {
+			observation.error(ex);
+			observation.stop();
+			throw ex;
+		}
+	}
+
+	private Observation newObservation(PulsarMessageSenderContext senderContext) {
+		if (this.observationRegistry == null) {
+			return Observation.NOOP;
+		}
+		return PulsarTemplateObservation.TEMPLATE_OBSERVATION.observation(this.observationConvention,
```

Comment:
Is it OK to reuse PulsarTemplateObservation ?
Or should the reactive template have its own ?

Reply:
As far as the bean names are concerned there should be no collision.
For the configuration, it would be under `spring.pulsar.template`. We currently don't have anything else that could differ between regular and reactive template config, but maybe one day we could ?

Reply:
> Is it OK to reuse PulsarTemplateObservation ?
> Or should the reactive template have its own Observation class ?

Excellent questions. I think using the same names w/ a distinguishing tag would work well. 

I would not imagine a single app instance using both reactive/non-reactive, but I could imagine a case where someone is doing a bake-off between the 2 and has 50% of their app instances using reactive and the other 50% non-reactive. It would be helpful be able to filter the metrics by client type etc..

All this being said, I would be in favor of pulling the O11Y out of the reactive side **for now**. The reason is it is a very new API and we may run into some kinks w/ it and I would prefer to get some runtime on it (and fix it in) in a single area before we proliferate it through the codebase. Wdyt?

Reply:
OK. I can remove the O11y.
Then the value of the Spring template compared to the raw client seems very thin...

Reply:
done